File: src/persist_sqlite.py
# ============================================================
# persist_sqlite.py — SQLite adapter for persist
#
# Usage:
#   from persist_sqlite import sqlite_adapter
#   from persist import persist
#   persist(engine, sqlite_adapter("state.db"))
#
# Or just:  persist(engine, "state.db")  — auto-detected
# ============================================================


def sqlite_adapter(path):
    """Create a SQLite persist adapter.

    path — file path or ":memory:"
    """
    import sqlite3
    conn = sqlite3.connect(path)
    conn.execute("PRAGMA journal_mode=WAL")

    return {
        "setup": lambda: (
            conn.execute("CREATE TABLE IF NOT EXISTS facts (term TEXT PRIMARY KEY)"),
            conn.commit()
        ),
        "insert": lambda key: conn.execute(
            "INSERT OR IGNORE INTO facts VALUES (?)", (key,)
        ),
        "remove": lambda key: conn.execute(
            "DELETE FROM facts WHERE term = ?", (key,)
        ),
        "all": lambda: [row[0] for row in conn.execute("SELECT term FROM facts")],
        "commit": lambda: conn.commit(),
        "close": lambda: conn.close(),
    }

File: src/persist-sqlite.js
// ============================================================
// persist-sqlite.js — SQLite adapter for persist
//
// Usage with better-sqlite3:
//   var Database = require('better-sqlite3');
//   persist(engine, sqliteAdapter(new Database('state.db')));
//
// Or with bun:sqlite:
//   var { Database } = require('bun:sqlite');
//   persist(engine, sqliteAdapter(new Database('state.db')));
//
// Or just:  persist(engine, db)  — auto-detected if db has .prepare
// ============================================================

function sqliteAdapter(db) {
  var cache = {};
  function stmt(sql) {
    if (!cache[sql]) cache[sql] = db.prepare(sql);
    return cache[sql];
  }

  return {
    setup: function() {
      db.exec("CREATE TABLE IF NOT EXISTS facts (term TEXT PRIMARY KEY)");
    },
    insert: function(key) {
      stmt("INSERT OR IGNORE INTO facts VALUES (?)").run(key);
    },
    remove: function(key) {
      stmt("DELETE FROM facts WHERE term = ?").run(key);
    },
    all: function() {
      return stmt("SELECT term FROM facts").all().map(function(r) { return r.term; });
    },
    commit: function() {},  // better-sqlite3 / bun:sqlite are autocommit
    close: function() { db.close(); }
  };
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.sqliteAdapter = sqliteAdapter;
}
export { sqliteAdapter };
